### ISSUE TITLE
Error in computing the size of spatial data when assigning from vector

### ISSUE THREAD

**Issue by jpswensen** [laspy/laspy#196]
I am updating my code from pylas to laspy and ran into what I think is a weird bug. I read through the migration guide, but couldn't see that my old code was doing something wrong. I made a super quick example that demonstrates the problem. 

I am more than willing to hunt down the problem and submit a pull request, but during a quick inspection of the various source files I can't easily find the problem. I figured I would ask here before diving in, in the case someone with  more expertise can either point me in the right direction, or if it is super easy to fix.

In the example below, I have a simple function that expects as input a matrix of dimensions Nx7. Each column represents a portion of my point cloud data. You can see how in the function I extract each column and assign it to some field of the LAZ file. The problem is that if each column is 1000 elements (like my example), it seems that laspy is internally creating a 999 element vector (really 1 element less than whatever the length of the vector is).

The error occurs at the line 
`las.x = colorizedPCL[:,0]`
and throws the exception
`ValueError: could not broadcast input array from shape (1000,) into shape (999,)`

The complete test file to demonstrate is below:
```Python
import laspy
import numpy as np

def writeLazFile(filename, timestamp, color_exposure, ir_exposure, colorizedPCL):
    
    # Create the las object and add extra dimensions for the exposure
    las = laspy.create(point_format=8, file_version='1.4')
    las.add_extra_dim(laspy.ExtraBytesParams(
        name="color_exposure",
        type="float64",
        description="Exposure of the color image"
    ))

    las.add_extra_dim(laspy.ExtraBytesParams(
        name="ir_exposure",
        type="float64",
        description="Exposure of the ir image"
    ))

    las.header.x_scale = 0.0001
    las.header.y_scale = 0.0001
    las.header.z_scale = 0.0001
    

    # Start populating the data
    las.gps_time = [timestamp]

    las.color_exposure = [color_exposure]
    las.ir_exposure = [ir_exposure]

    las.x = colorizedPCL[:,0]
    las.y = colorizedPCL[:,1]
    las.z = colorizedPCL[:,2]

    las.red = np.uint16(colorizedPCL[:,3])
    las.green = np.uint16(colorizedPCL[:,4]) 
    las.blue = np.uint16(colorizedPCL[:,5]) 

    las.intensity = colorizedPCL[:,6]

    las.write(filename)

filename = 'test.laz'
timestamp = '2022'
color_exposure = 1.0
ir_exposure = 1.0
colorizedPCL = np.zeros((1000,7))
writeLazFile(filename, timestamp, color_exposure, ir_exposure, colorizedPCL)
```

**Comment by tmontaigu** [comment]
The problem is both in your code and laspy (albeit its a tiny bit in your code and mostly in laspy).

What happens is that you do:
```Python
    # Start populating the data
    las.gps_time = [timestamp]

    las.color_exposure = [color_exposure]
    las.ir_exposure = [ir_exposure]
```

So you are settings, `gps_time`, `color_exposure`, `ir_exposure` to array of length 1. 
So the internal array used by laspy now has a length of 1.

And later you do `las.x = colorizedPCL[:,0]`. 
And this is where the laspy bug triggers, as internally, instead of resizing the array to have a length of 1000
we resized it to have a length of 999 (the difference between current size (1) and input size (1000)).
So just after the wrong resizing we tried to set the value, but it triggered the broadcast error since the length was not correct.

The PR #198 will fix this.

In your code, to avoid to problem you could directly set the arrays to 1000 elements:

```Python
    # Start populating the data
    las.gps_time = [timestamp] * colorizedPCL.shape[0]
    las.color_exposure = [color_exposure] * colorizedPCL.shape[0]
    las.ir_exposure = [ir_exposure] * colorizedPCL.shape[0]
```

**Comment by jpswensen** [comment]
After I saw your answer, I went back and realized I had misunderstood
completely extra dimensions. I thought they could be a different size than
the [X,Y,Z,R,G,B] data. Since we are using a camera instead of something
like a lidar, our capture time and exposure times are really the same for
all pixels and we feel it would be a waste of disk space to make those
extra dimensions. So, I figured out how to do VLRs and made each of those
three values a scalar-valued VLR.

Thanks for the quick response. I will mark this closed, if you have no
objections.
